### PR TITLE
Extensions: do not trigger unnecessary getLocations calls

### PR DESCRIPTION
**Pull request by chrismwendt**
Previously, `getLocations` would be called every time an extension would set properties on the panel view.

Now it only gets called when the `component` changes.

### CODE REVIEW

**Comment by chrismwendt** [comment]
I merged this in quick for dotGo.

File: shared/src/api/client/api/views.ts [hunk 1]
```diff
@@ -37,21 +37,32 @@ export class ClientViews implements ClientViewsAPI {
...
-                          )
-                        : undefined
+            combineLatest(
+                panelView.pipe(
+                    map(({ title, content, priority }) => ({ title, content, priority })),
+                    distinctUntilChanged((x, y) => isEqual(x, y))
+                ),
+                panelView.pipe(
+                    map(({ component }) => component),
+                    filter((component): component is { locationProvider: string } => Boolean(component)),
```
Comment (lguychard) [inline shared/src/api/client/api/views.ts]: You can use `filter(isDefined)`

File: shared/src/api/client/api/views.ts [hunk 2]
```diff
@@ -37,21 +37,32 @@ export class ClientViews implements ClientViewsAPI {
...
+                    map(({ locationProvider }) => locationProvider),
+                    distinctUntilChanged(),
+                    map(locationProvider =>
+                        from(this.modelService.model).pipe(
+                            switchMap(model => {
+                                const params: TextDocumentPositionParams | null = modelToTextDocumentPositionParams(
```
Comment (lguychard) [inline shared/src/api/client/api/views.ts]: Why do you need to declare it as `TextDocumentPositionParams | null`? Isn't the return type inferred correctly?

File: shared/src/api/client/api/views.ts [hunk 3]
```diff
@@ -37,21 +37,32 @@ export class ClientViews implements ClientViewsAPI {
...
-                                  if (!params) {
-                                      return of(of(null))
-                                  }
-                                  return this.textDocumentLocations.getLocations(component.locationProvider, params)
-                              })
-                          )
-                        : undefined
+            combineLatest(
+                panelView.pipe(
+                    map(({ title, content, priority }) => ({ title, content, priority })),
```
Comment (lguychard) [inline shared/src/api/client/api/views.ts]: This had me pause for a second - I find its purpose (excluding `component`) unclear. Maybe use `_.omit`?

```typescript
map(panelViewData => omit(panelViewData, 'component'))
```

**Comment by chrismwendt** [comment]
Thanks @lguychard, I'm addressing all of your feedback in https://github.com/sourcegraph/sourcegraph/pull/3000

**Pull request merged**